### PR TITLE
8316422: TestIntegerUnsignedDivMod.java triggers "invalid layout" assert in FrameValues::validate

### PR DESCRIPTION
**Pull request by dean-long**
Don't truncate caller exception state stack in c1, even if no handlers are found.  This makes c1 follow the example of c2.  See bug for details.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316422](https://bugs.openjdk.org/browse/JDK-8316422): TestIntegerUnsignedDivMod.java triggers "invalid layout" assert in FrameValues::validate (**Bug** - P4)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [b7a51cf9](https://git.openjdk.org/jdk/pull/16297/files/b7a51cf900a8c39aa3e7e9f1b38daa47580e6716)
 * [Tom Rodriguez](https://openjdk.org/census#never) (@tkrodriguez - **Reviewer**) ⚠️ Review applies to [e54fa4c0](https://git.openjdk.org/jdk/pull/16297/files/e54fa4c0370ebcc3a7e3995311311ebd6faf73b8)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16297/head:pull/16297` \
`$ git checkout pull/16297`

Update a local copy of the PR: \
`$ git checkout pull/16297` \
`$ git pull https://git.openjdk.org/jdk.git pull/16297/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16297`

View PR using the GUI difftool: \
`$ git pr show -t 16297`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16297.diff">https://git.openjdk.org/jdk/pull/16297.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16297#issuecomment-1773449996)

### CODE REVIEW

**Comment by dean-long** [comment]
Some details.  When throwing an exception, the top frame pops everything from the stack before pushing the exception object, so the exception state recorded for deoptimization does the same.  However, c1 was incorrectly emptying the stack for caller frames as well.  Now, instead of emptying the stack for the caller, we mark those slots invalid.
For locals, c1 front-end was also removing all the locals, then adding them back before generating the deopt info.  Now locals are invalidated like the stack if there is not exception handler.
Finally, local values need to be retained instead of invalidated when JVMTI is enabled.  I refactored the code to simplify this logic.
Some iterators needed to be changed to support nullptr values in the stack state.  This was already supported for locals, so now stack/local treatment is more uniform.

**Comment by dean-long** [comment]
I got some test failures, so moving back to draft.

**Comment by TobiHartmann** [comment]
I think a regression test would be good as well.

File: test/hotspot/jtreg/compiler/exceptions/TestDeoptExceptionState.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+* @test
+* @bug 8316422
+* @summary Test exception state used for deoptimization.
+* @run main/othervm -XX:+VerifyStack -Xcomp -XX:TieredStopAtLevel=1 -XX:+DeoptimizeALot
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/exceptions/TestDeoptExceptionState.java:28]: `VerifyStack` and `DeoptimizeALot` are debug flags.

File: src/hotspot/share/c1/c1_ValueStack.hpp [hunk 2]
```diff
@@ -36,6 +36,8 @@ class ValueStack: public CompilationResourceObj {
     StateAfter,          // After execution of instruction
     ExceptionState,      // Exception handling of instruction
     EmptyExceptionState, // Exception handling of instructions not covered by an xhandler
+    CallerExceptionState,
+    CallerEmptyExceptionState,
```
Comment (tkrodriguez) [inline src/hotspot/share/c1/c1_ValueStack.hpp:40]: Maybe add a comment for these new kinds

**Review by tkrodriguez — APPROVED** [review]
Looks ok to me.

File: src/hotspot/share/c1/c1_ValueStack.hpp (hunk 2 above)
Comment (dean-long) [reply src/hotspot/share/c1/c1_ValueStack.hpp]: Done.

**Comment by dean-long** [comment]
Thanks @TobiHartmann and @tkrodriguez .

**Pull request closed without merging**